src/Control/Effect/Empty.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE DeriveFunctor, DeriveGeneric, FlexibleContexts, FlexibleInstances, KindSignatures, MultiParamTypeClasses, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Empty
+( -- * Empty effect
+  Empty(..)
+, abort
+  -- * Empty carrier
+, runEmpty
+, EmptyC(..)
+  -- * Re-exports
+, Carrier
+, Member
+, run
+) where
+
+import Control.Applicative (Alternative (..), liftA2)
+import Control.Effect.Carrier
+import Control.Monad (MonadPlus (..))
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import GHC.Generics (Generic1)
+
+data Empty (m :: * -> *) k = Empty
+  deriving (Functor, Generic1)
+
+instance HFunctor Empty
+instance Effect   Empty
+
+-- | Abort the computation.
+--
+--   prop> run (runEmpty abort) === Nothing
+abort :: (Carrier sig m, Member Empty sig) => m a
+abort = send Empty
+
+
+-- | Run an 'Empty' effect, returning 'Nothing' for empty computations, or 'Just' the result otherwise.
+--
+--   prop> run (runEmpty abort)    === Nothing
+--   prop> run (runEmpty (pure a)) === Just a
+runEmpty :: EmptyC m a -> m (Maybe a)
+runEmpty = runEmptyC
+
+newtype EmptyC m a = EmptyC { runEmptyC :: m (Maybe a) }
+  deriving (Functor)
+
+instance Applicative m => Applicative (EmptyC m) where
+  pure = EmptyC . pure . Just
+  EmptyC f <*> EmptyC a = EmptyC (liftA2 (<*>) f a)
```

Comment:
I think we should probably INLINE these definitions, as `MaybeT` from `transformers` does.